forms/submission_form.py

```diff
@@ -0,0 +1,165 @@
+from uuid import uuid4
+import copy
+from typing import List, Tuple
+
+import sqlalchemy as sql
+from sqlalchemy.orm.attributes import flag_modified
+
+from models import (
+    Taxon,
+    Metabolite,
+    Project,
+    Study,
+    Submission,
+)
+
+
+class SubmissionForm:
+    def __init__(self, submission_id, step, db_session=None, user_uuid=None):
+        self.step       = step
+        self.db_session = db_session
+
+        # Load submission object
+        self.submission = None
+        if submission_id is not None:
+            self.submission = self.db_session.get(Submission, submission_id)
+
+        if self.submission is None:
+            self.submission = Submission(
+                projectUniqueID=None,
+                studyUniqueID=None,
+                userUniqueID=user_uuid,
+                studyDesign={
+                    'project': {
+                        'name':        None,
+                        'description': None,
+                    },
+                    'vessel_type':     None,
+                    'bottle_count':    None,
+                    'plate_count':     None,
+                    'vessel_count':    None,
+                    'column_count':    None,
+                    'row_count':       None,
+                    'timepoint_count': None,
+                    'technique_types': [],
+                    'strains':         [],
+                    'new_strains':     [],
+                    'metabolites':     [],
+                }
+
+            )
+
+        # Check for an existing project/study and set the submission "type" accordingly:
+        self.project_id = self._find_project_id()
+        self.study_id   = self._find_study_id()
+        self.type       = self._determine_project_type()
+
+    def update_project(self, data):
+        self.type = data['submission_type']
+        self.submission.studyDesign['project'] = {
+            'name':        data['project_name'],
+            'description': data['project_description'],
+        }
+        flag_modified(self.submission, 'studyDesign')
+
+        if self.type == 'new_project':
+            self.submission.projectUniqueID = str(uuid4())
+            self.submission.studyUniqueID   = str(uuid4())
+        elif self.type == 'new_study':
+            self.submission.projectUniqueID = data['project_uuid']
+            self.submission.studyUniqueID   = str(uuid4())
+        elif self.type == 'update_study':
+            self.submission.projectUniqueID = data['project_uuid']
+            self.submission.studyUniqueID   = data['study_uuid']
+
+    def update_strains(self, data):
+        self.submission.studyDesign['strains']     = data['strains']
+        self.submission.studyDesign['new_strains'] = data['new_strains']
+        flag_modified(self.submission, 'studyDesign')
```

Comment:
Because the data is stored in a JSON field that is stored as TEXT, it seems SQLAlchemy does not automatically keep track of modifications to it for performance reasons. We need to manually flag this field as modified for it to know to trigger an update on `db_session.commit()`.

Reply:
Oh! I think you documented here already the `flag_modified` !